wrappers/python/tests/TestModeller.py

```diff
@@ -593,6 +593,40 @@ def test_addHydrogensPdb3(self):
 
         validate_equivalence(self, topology_start, topology_after)
 
+    def test_addHydrogensPdb3_keepPositions(self):
+        """ Test addHydrogens() does not change existing Hs positions """
+
+        # build the Modeller
+        topology_start = self.topology_start3
+        positions = self.positions3.value_in_unit(nanometers)
+        modeller = Modeller(topology_start, positions)
+
+        # Remove some hydrogens from the structure
+        res_list = list(topology_start.residues())
+        toDelete = [atom for atom in res_list[-1].atoms() if atom.element == element.hydrogen]
+        n_deleted = len(toDelete)
+        modeller.delete(toDelete)
+
+        # record positions of hydrogens before calling addHydrogens
+        oriH = [atom.index for atom in modeller.topology.atoms() if atom.element == element.hydrogen]
+        oriH_pos = [positions[i] for i in oriH]
+
+        # add the hydrogens back
+        modeller.addHydrogens(self.forcefield)
+        topology_after = modeller.getTopology()
+
+        # Fetch 'new' positions
+        new_positions = modeller.positions.value_in_unit(nanometers)
+        newH = [atom.index for atom in topology_after.atoms() if atom.element == element.hydrogen]
+        newH_pos = [new_positions[i] for i in newH][:-1*n_deleted]
+
+        # added all Hs back in?
+        self.assertEqual(len(newH), len(oriH) + n_deleted)
+
+        # Are the old ones at the same position?
+        xyz_diff = any([sum(o-n) for o, n in zip(oriH_pos, newH_pos)])
```

Comment:
I'd change this to

```python
xyz_diff = any([norm(o-n) > 1e-6 for o, n in zip(oriH_pos, newH_pos)])
```

The positions will have been passed to a Context and back again, so depending on the platform it's possible they'll have been converted to single precision along the way.  Also, I prefer to avoid implicit conversions of other types to booleans, since they're confusing to read and sometimes lead to unexpected results.

Reply:
Good point. My first thought was to use the `numpy.isclose` function but I didn't want to import numpy, convert all positions, etc, just for this. I'll use the norm function then and that tolerance factor.